File: brain-move/games/memory_immagini.py
import random
import time
from games.base_game import BaseGame


class MemoryImmagini(BaseGame):
    FAMILIAR = {
        "nonno": "\U0001F474",
        "nonna": "\U0001F475",
        "mamma": "\U0001F469",
        "papa": "\U0001F468",
        "bambino": "\U0001F476",
        "bambina": "\U0001F467",
        "casa": "\U0001F3E0",
        "macchina": "\U0001F697",
        "cane": "\U0001F436",
        "gatto": "\U0001F431",
        "albero": "\U0001F333",
        "fiore": "\U0001F33C",
        "pane": "\U0001F35E",
        "mela": "\U0001F34E",
        "acqua": "\U0001F4A7",
        "sole": "\U0001F31E",
        "luna": "\U0001F319",
        "stella": "\u2B50",
        "cuore": "\u2764\uFE0F",
        "sorriso": "\U0001F60A",
        "libro": "\U0001F4DA",
        "palla": "\u26BD",
        "torta": "\U0001F370",
        "orologio": "\U0001F570",
    }

    def __init__(self, difficulty_engine, scoring_engine):
        super().__init__(difficulty_engine, scoring_engine)
        self.name = "memory_immagini"
        self.display_name = "Memory Immagini"
        self.phase = "instruction"
        self.grid_rows = 0
        self.grid_cols = 0
        self.cards = []
        self.revealed = []
        self.matched = []
        self.first_card = None
        self.waiting = False
        self.wait_timer = 0
        self.pairs_found = 0
        self.total_pairs = 0
        self.attempts = 0
        self.instruction_timer = 0
        self.study_phase = False
        self.study_timer = 0
        self.study_duration = 5

    def start(self):
        super().start()
        level = self.difficulty.get_level()
        self.grid_cols = min(3 + level, 6)
        self.grid_rows = min(2 + level // 2, 4)
        if self.grid_cols * self.grid_rows % 2 != 0:
            self.grid_rows += 1
        self.total_pairs = (self.grid_cols * self.grid_rows) // 2
        self.study_duration = max(3, 7 - level)
        self.phase = "instruction"
        self.instruction_timer = time.time()
        self.feedback_message = f"Memorizza le immagini! Hai {self.study_duration} secondi"

        keys = list(self.FAMILIAR.keys())
        random.shuffle(keys)
        chosen = keys[:self.total_pairs]
        deck = chosen * 2
        random.shuffle(deck)
        self.cards = []
        for label in deck:
            self.cards.append({"label": label, "icon": self.FAMILIAR[label]})
        self.revealed = [False] * len(self.cards)
        self.matched = [False] * len(self.cards)
        self.first_card = None
        self.waiting = False
        self.pairs_found = 0
        self.attempts = 0

    def update(self, pose_data=None, dt=0):
        if not self.running:
            return

        if self.phase == "instruction":
            if time.time() - self.instruction_timer > 2.0:
                self.phase = "study"
                self.study_timer = time.time()
                self.study_phase = True
                for i in range(len(self.cards)):
                    self.revealed[i] = True

        elif self.phase == "study":
            if time.time() - self.study_timer > self.study_duration:
                self.phase = "playing"
                for i in range(len(self.cards)):
                    self.revealed[i] = False
                self.feedback_message = "Trova le coppie!"

        elif self.phase == "playing" and self.waiting:
            self.wait_timer -= dt
            if self.wait_timer <= 0:
                self._end_wait()

    def select_card(self, idx):
        if self.phase != "playing" or self.waiting:
            return False
        if self.matched[idx] or self.revealed[idx]:
            return False

        self.revealed[idx] = True
        if self.first_card is None:
            self.first_card = idx
            return True

        self.attempts += 1
        second = idx
        if self.cards[self.first_card]["label"] == self.cards[second]["label"]:
            self.matched[self.first_card] = True
            self.matched[second] = True
            self.pairs_found += 1
            self.first_card = None
            self.feedback_message = f"Trovata coppia! {self.pairs_found}/{self.total_pairs}"
            if self.pairs_found == self.total_pairs:
                self._finish_game()
            return True
        else:
            self.waiting = True
            self.wait_timer = 1.2
            self._second_card = second
            self.feedback_message = "Non e' la coppia giusta"
            return True

    def _end_wait(self):
        self.revealed[self.first_card] = False
        self.revealed[self._second_card] = False
        self.first_card = None
        self.waiting = False

    def _finish_game(self):
        self.stop()
        accuracy = self.total_pairs / max(self.attempts, 1)
        self.result = {
            "completed": True,
            "accuracy": accuracy,
            "pairs_found": self.pairs_found,
            "attempts": self.attempts,
            "grid": f"{self.grid_rows}x{self.grid_cols}",
        }
        self.difficulty.adapt(accuracy)
        self.scoring.add_score(self.name, int(accuracy * 30))

    def get_progress(self):
        return self.pairs_found / max(self.total_pairs, 1)

    def is_card_revealed(self, idx):
        return self.revealed[idx]

    def is_card_matched(self, idx):
        return self.matched[idx]

    def get_card_icon(self, idx):
        if 0 <= idx < len(self.cards):
            return self.cards[idx]["icon"]
        return "?"

File: brain-move-web/src/games/memory-immagini.ts
import { BaseGame } from './base-game'

const FAMILIAR: Record<string, string> = {
  nonno: "\u{1F474}", nonna: "\u{1F475}", mamma: "\u{1F469}", papa: "\u{1F468}",
  bambino: "\u{1F476}", bambina: "\u{1F467}", casa: "\u{1F3E0}", macchina: "\u{1F697}",
  cane: "\u{1F436}", gatto: "\u{1F431}", albero: "\u{1F333}", fiore: "\u{1F33C}",
  pane: "\u{1F35E}", mela: "\u{1F34E}", acqua: "\u{1F4A7}", sole: "\u{1F31E}",
  luna: "\u{1F319}", stella: "\u2B50", cuore: "\u2764\uFE0F", sorriso: "\u{1F60A}",
  libro: "\u{1F4DA}", palla: "\u26BD", torta: "\u{1F370}", orologio: "\u{1F570}",
}

export class MemoryImmagini extends BaseGame {
  phase: "instruction" | "study" | "playing" = "instruction"
  gridRows = 0
  gridCols = 0
  cards: Array<{ label: string; icon: string }> = []
  revealed: boolean[] = []
  matched: boolean[] = []
  firstCard: number | null = null
  secondCardIdx: number | null = null
  waiting = false
  waitTimer = 0
  pairsFound = 0
  totalPairs = 0
  attempts = 0
  instructionTimer = 0
  studyTimer = 0
  studyDuration = 5

  constructor(difficulty: any, scoring: any) {
    super(difficulty, scoring)
    this.name = "memory_immagini"
    this.displayName = "Memory Immagini"
  }

  start(): void {
    super.start()
    const level = this.difficulty.getLevel()
    this.gridCols = Math.min(3 + level, 6)
    this.gridRows = Math.min(2 + Math.floor(level / 2), 4)
    if (this.gridCols * this.gridRows % 2 !== 0) this.gridRows++
    this.totalPairs = (this.gridCols * this.gridRows) / 2
    this.studyDuration = Math.max(3, 7 - level)
    this.phase = "instruction"
    this.instructionTimer = Date.now()
    this.feedbackMessage = `Memorizza le immagini! Hai ${this.studyDuration} secondi`
    const keys = Object.keys(FAMILIAR).sort(() => Math.random() - 0.5).slice(0, this.totalPairs)
    const deck = [...keys, ...keys].sort(() => Math.random() - 0.5)
    this.cards = deck.map(label => ({ label, icon: FAMILIAR[label] }))
    this.revealed = new Array(this.cards.length).fill(false)
    this.matched = new Array(this.cards.length).fill(false)
    this.firstCard = null
    this.waiting = false
    this.pairsFound = 0
    this.attempts = 0
  }

  update(_poseData?: any, dt = 0): void {
    if (!this.running) return
    if (this.phase === "instruction" && (Date.now() - this.instructionTimer) / 1000 > 2) {
      this.phase = "study"
      this.studyTimer = Date.now()
      this.revealed = new Array(this.cards.length).fill(true)
    } else if (this.phase === "study" && (Date.now() - this.studyTimer) / 1000 > this.studyDuration) {
      this.phase = "playing"
      this.revealed = new Array(this.cards.length).fill(false)
      this.feedbackMessage = "Trova le coppie!"
    } else if (this.phase === "playing" && this.waiting) {
      this.waitTimer -= dt
      if (this.waitTimer <= 0) this._endWait()
    }
  }

  selectCard(idx: number): boolean {
    if (this.phase !== "playing" || this.waiting || this.matched[idx] || this.revealed[idx]) return false
    this.revealed[idx] = true
    if (this.firstCard === null) { this.firstCard = idx; return true }
    this.attempts++
    const second = idx
    if (this.cards[this.firstCard].label === this.cards[second].label) {
      this.matched[this.firstCard] = true; this.matched[second] = true
      this.pairsFound++
      this.firstCard = null
      this.feedbackMessage = `Trovata coppia! ${this.pairsFound}/${this.totalPairs}`
      if (this.pairsFound === this.totalPairs) this._finishGame()
      return true
    }
    this.waiting = true
    this.waitTimer = 1.2
    this.secondCardIdx = second
    this.feedbackMessage = "Non è la coppia giusta"
    return true
  }

  private _endWait(): void {
    if (this.firstCard !== null) this.revealed[this.firstCard] = false
    if (this.secondCardIdx !== null) this.revealed[this.secondCardIdx] = false
    this.firstCard = null; this.waiting = false
  }

  private _finishGame(): void {
    this.stop()
    const accuracy = this.totalPairs / Math.max(this.attempts, 1)
    this.result = { completed: true, accuracy: Math.min(accuracy, 1), pairsFound: this.pairsFound, attempts: this.attempts }
    this.difficulty.adapt(Math.min(accuracy, 1))
    this.scoring.addScore(this.name, Math.round(accuracy * 30))
  }

  getProgress(): number { return this.pairsFound / Math.max(this.totalPairs, 1) }
  isCardRevealed(idx: number): boolean { return this.revealed[idx] }
  isCardMatched(idx: number): boolean { return this.matched[idx] }
  getCardIcon(idx: number): string { return this.cards[idx]?.icon ?? "?" }
  getGridSize(): [number, number] { return [this.gridRows, this.gridCols] }
}
